### ISSUE TITLE
loadbalancing on sequential runs changes the grid

### ISSUE THREAD

**Issue by andlaus** [OPM/opm-grid#197]
if I create the grid from the Norne deck from opm-data with MPI enabled but in sequential mode, the number of elements in the first (and only) grid partition is different before and after the call to `loadBalance()`. This can be tested using the following program:

```c++
#include "config.h"

#include <opm/parser/eclipse/Parser/Parser.hpp>
#include <opm/parser/eclipse/Parser/ParseMode.hpp>
#include <opm/parser/eclipse/Deck/Deck.hpp>
#include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
#include <dune/grid/CpGrid.hpp>

#include <dune/common/parallel/mpihelper.hh>

int main(int argc, char **argv)
{
    Dune::MPIHelper::instance(argc, argv);

    Opm::Parser parser;
    Opm::ParseMode parseMode({{ Opm::ParseMode::PARSE_RANDOM_SLASH , Opm::InputError::IGNORE }});
    const auto deck = parser.parseFile(argv[1], parseMode);
    const Opm::EclipseState eclState(deck, parseMode);

    std::vector<double> porv = eclState.getDoubleGridProperty("PORV")->getData();

    Dune::CpGrid grid;
    grid.processEclipseFormat(eclState.getEclipseGrid(),
                              /*isPeriodic=*/false,
                              /*flipNormals=*/false,
                              /*clipZ=*/false,
                              porv);


    int elemsBefore = grid.leafGridView().size(0);

    grid.loadBalance();
    grid.switchToDistributedView();

    int elemsAfter = grid.leafGridView().size(0);
    std::cout << "elemsBefore: " << elemsBefore << "\n"
              << "elemsAfter: " << elemsAfter << "\n";
    if (elemsBefore != elemsAfter)
        std::cout << "oops: number of elements is different!\n";

    return 0;
}
```

which on my machine produces the following output when it is feed with the Norne deck:

```
and@heuristix:~/tmp > ./cpTest /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
VFPPROD bhp versus thp not monotonic increasing: 7/4224(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 9/3168(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 1/7296(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 6/15200(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 5/13300(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 11/13300(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 8/13300(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 10/13300(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 15/13680(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 3/15200(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 73/3360(2%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 335/6804(4%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 32/6804(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 54/6804(0%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 282/6912(4%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 87/6804(1%) elements failed test
VFPPROD bhp versus thp not monotonic increasing: 337/6048(5%) elements failed test
ERROR when handling keyword for well C-4H: Cannot open a well where all completions are shut
ERROR when handling keyword for well D-1H: Cannot open a well where all completions are shut
ERROR when handling keyword for well D-1H: Cannot open a well where all completions are shut
ERROR when handling keyword for well E-3AH: Cannot open a well where all completions are shut
ERROR when handling keyword for well D-3AH: Cannot open a well where all completions are shut
ERROR when handling keyword for well E-2H: Cannot open a well where all completions are shut
After loadbalancing process 0 has 44420 cells.
elemsBefore: 44431
elemsAfter: 44420
oops: number of elements is different!
```

**Comment by atgeirr** [comment]
A possible reason (not checked): the Norne grid consists of 2 disjoint parts, I do not recall exactly how many cells the smaller one has, but it is around the number missing here. If the partitioner uses connectivity (grid, nnc and well) to build the partitions, parts that are totally disjoint would not be included.

**Comment by andlaus** [comment]
hm, sounds like grid-wise the Norne deck is buggy. Anyway, disconnected parts of the grid should not be thrown away at load balancing.

**Comment by blattms** [comment]
On Wed, Jan 06, 2016 at 03:37:31AM -0800, Atgeirr Flø Rasmussen wrote:
> A possible reason (not checked): the Norne grid consists of 2
> disjoint parts.

I think that this is the right clue. If one is not using Zoltan (which
I silenly assume here), then the methods in
`dune/grid/common/GridPartitioning.cpp` are used. I must admit that I
always assumed that they are doing the right thing. Unfortunately, I
am not the author. But it seems like the method `ensureConnectivity`
might result in more partitions than requested (e.g. in a serial run 2
could be produced). But this change in the number of partitions is not
yet honored by the code.

@andlaus Are you using Zoltan or not?

**Comment by andlaus** [comment]
> @andlaus Are you using Zoltan or not?

not ;)

seems like these are actually to separate issues: one is the Norne deck (I think that the disconnected grid partition is not supposed to be there), the other is that dune-cornerpoint should handle this anyway...

**Comment by andlaus** [comment]
I'm pretty sure that this issue still applies. are there any plans to fix it? I suppose fixing it for the 2016.04 is too ambitious?

**Comment by atgeirr** [comment]
To clarify this: when using Zoltan, will the disjoint cells be part of a single partition in the single-process case?

If so, and the bug only manifests when using MPI without Zoltan, I don't think we are in a hurry to fix it since the Zoltan-using version is the only one I consider to have a future. I think that at some point (but maybe not before the release) we might want to make Zoltan mandatory if you want flow_mpi, however those libs are nontrivial to install correctly, so I hesitate to make it so.

**Comment by blattms** [comment]
Yep, with ZOLTAN the problem is not present.

**Comment by andlaus** [comment]
fixed by #245. closing
